docs/design/database.md

```diff
@@ -0,0 +1,129 @@
+        string hub_router
+        string hub_zone
+        string zone_prefix
+        string reflexive_ipv4
+        string endpoint_local_address_ipv4
+        string symmetric_nat
+    }
+```
+
+We have 4 concepts:
+
+- Users: Unique users of the service
+- Devices: A unique end-user device
+- Zone: A way of describing which devices may communicate with eachother
+- Peer: An association between a Device and a Zone. This is used by the apex agent to identify who it may peer with.
+
+Since connectivity was Zone-based, we would create a "Default Zone" where all devices would be onboarded into. Adding > 1 Zone was disabled, since there were technical challenges around moving a device from one Zone to another, or for a device to have peers in > 1 Zone.
+
+The device-onboarding flow was as follows:
+
+1. Operator/Customer sets up OIDC to provide users with access to Apex
+1. User runs the Apex agent
+1. Apex agent does a `POST /device` to register the unique device with Apex for the user. It is assigned into the single zone that a user is in (either the default zone, or a new zone where the user was moved to by an admin).
+1. Once the device is registered a `POST /zones/$id/peers` was issued to create an association between this device and a zone.
+1. `GET /zones/$id/peers` was polled to retrieve the peer listing for a zone.
+
+On top of this flow, we were somewhat abusing `POST /zones/$id/peers` to also update peers after NAT-T.
+
+This model was built under the assumption that multi-tenancy would be achieved by deploying an Apex stack per-tenant.
+
+## v2: SaaS Preview Milestone
+
+See [issue #348](https://github.com/redhat-et/apex/issues/348).
+
+In order to create a more SaaS-ready schema, we've made the following changes.
+
+1. Replace the `Zone` construct with `Organization`. This is a GitHub-like construct, where each unique `User` is also its own `Organization`.
+1. Deprecate `Peer` and combine it's functionality with the `Device` table.
+
+```mermaid
+erDiagram
+    USER ||--o{ DEVICE : has
+    USER ||--|{ ORGANIZATION : belongs_to
+    ORGANIZATION ||--o{ DEVICE: contains
+    USER{
+        string id
+        string username
+    }
+    DEVICE{
+        string id
+        string user_id
+        string organization_id
+        string public_key
+        string hostname
+        string tunnel_ip
+        string allowed_ips
+        string local_ip
+        string child_prefix
+        string relay
+        string organization_prefix
+        string reflexive_ipv4
+        string endpoint_local_address_ipv4
+        string symmetric_nat
+    }
+    ORGANIZATION{
+        string id
+        string name
+        string description
+        string cidr
+        string hub_zone
+    }
+```
+
+In this simplified model we only have 3 concepts:
+
+- Users: Unique users of the service
+- Devices: A unique end-user device
+- Organization: A collection of users who have permissions to onboard one or more devices into this organization
```

Comment:
I think we need some clarity around users <-> organization relationship. Looks like we support both the models N:1 and 1:N for user and organization. 1:N relation is clear. But needs some clarification around N:1 model. 

- Every user will have it's own personal organization, and given that each organization can have multiple users, I assume that the original owner of the organization need to invite other users to his organization? Will that invited user have similar kind of ownership on the invited organization (e.g can this invited user invite other users)? or we will have a concept of original owner and invitees, where invitee only have "read" only access? I assume it's the later, but want to clarify.
- In v2 we are not allowing users to join multiple organization, but do we allow user to join any other organization (1:1) ?

Reply:
The diagram shows that a user may belong to 1 or more organizations - the first organization they belong to is their personal org.

> Every user will have it's own personal organization, and given that each organization can have multiple users, I assume that the original owner of the organization need to invite other users to his organization? Will that invited user have similar kind of ownership on the invited organization (e.g can this invited user invite other users)? or we will have a concept of original owner and invitees, where invitee only have "read" only access? I assume it's the later, but want to clarify.

The invite mechanism hasn't been designed yet - and is somewhat tied in with the ACL work we need to do across the API... but your assumption is correct. We'd have different permission levels in an org and the default would be read-only for everyone other than the owner.

> In v2 we are not allowing users to join multiple organization, but do we allow user to join any other organization (1:1) ?

I think this is backwards. V1 was 1:1 User:Zone. V2 is 1:Many(1 or more) User:Organization.
V2 can support joining multiple orgs, even if it takes a while for us to implement it. The main thing is the database layout wouldn't need to change

Reply:
Sounds good.